### PR TITLE
[spirv] Support TextureBuffer/ConstantBuffer arrays.

### PR DESCRIPTION
**Pull request by ehsannas**

### CODE REVIEW

File: tools/clang/lib/SPIRV/DeclResultIdMapper.cpp [hunk 1]
```diff
@@ -598,14 +616,15 @@ uint32_t DeclResultIdMapper::createCTBuffer(const VarDecl *decl) {
 }
 
 uint32_t DeclResultIdMapper::createPushConstant(const VarDecl *decl) {
+  // The front-end errors out if non-struct type push constant is used.
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply tools/clang/lib/SPIRV/DeclResultIdMapper.cpp]: We've already added the check in the front-end.

File: tools/clang/lib/SPIRV/TypeTranslator.cpp [hunk 2]
```diff
@@ -594,7 +594,14 @@ uint32_t TypeTranslator::translateType(QualType type, LayoutRule rule) {
   }
 
   if (const auto *arrayType = astContext.getAsConstantArrayType(type)) {
-    const uint32_t elemType = translateType(arrayType->getElementType(), rule);
+    const auto arrayElemType = arrayType->getElementType();
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply tools/clang/lib/SPIRV/TypeTranslator.cpp]: Yes.
+1 for having source location in TypeTranslator.

File: tools/clang/lib/SPIRV/TypeTranslator.cpp [hunk 3]
```diff
@@ -594,7 +594,14 @@ uint32_t TypeTranslator::translateType(QualType type, LayoutRule rule) {
...
+    if (isStructuredBuffer(arrayElemType) ||
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply tools/clang/lib/SPIRV/TypeTranslator.cpp]: Done

File: tools/clang/lib/SPIRV/TypeTranslator.cpp [hunk 4]
```diff
@@ -594,7 +594,14 @@ uint32_t TypeTranslator::translateType(QualType type, LayoutRule rule) {
...
+        isRWAppendConsumeSBuffer(arrayElemType)) {
+      emitError("using arrays of structured buffers is not allowed");
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply tools/clang/lib/SPIRV/TypeTranslator.cpp]: Done

File: tools/clang/test/CodeGenSPIRV/type.structured-buffer.array.error.hlsl [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+struct T {
+  float  a;
+  float3 b;
+};
+
+StructuredBuffer<T> myStructuredBuffer[3];
+RWStructuredBuffer<T> myRWStructuredBuffer[4];
+AppendStructuredBuffer<T> myAppendStructuredBuffer[2];
+ConsumeStructuredBuffer<T> myConsumeStructuredBuffer[2];
+
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply tools/clang/test/CodeGenSPIRV/type.structured-buffer.array.error.hlsl]: Done

**Comment by ehsannas** [comment]
Reached the 1-hour time limit.

File: tools/clang/lib/SPIRV/DeclResultIdMapper.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (antiagainst) [reply tools/clang/lib/SPIRV/DeclResultIdMapper.cpp]: Yep. Sorry I forgot about that, although I added that check. LoL

**Pull request merged**